packages/router/src/active-route-loader.tsx

```diff
@@ -0,0 +1,260 @@
+import React, { useRef, useState, useContext, useEffect } from 'react'
+
+import { unstable_batchedUpdates } from 'react-dom'
+
+import { Spec } from './util'
+import {
+  createNamedContext,
+  getAnnouncement,
+  getFocus,
+  resetFocus,
+} from './util'
+
+import { ParamsProvider, useLocation } from '.'
+
+const DEFAULT_PAGE_LOADING_DELAY = 1000 // milliseconds
+
+export interface PageLoadingContextInterface {
+  loading: boolean
+}
+
+export const PageLoadingContext =
+  createNamedContext<PageLoadingContextInterface>('PageLoading')
+
+export const usePageLoadingContext = () => {
+  const pageLoadingContext = useContext(PageLoadingContext)
+
+  if (!pageLoadingContext) {
+    throw new Error(
+      'usePageLoadingContext must be used within a PageLoadingContext provider'
+    )
+  }
+
+  return pageLoadingContext
+}
+
+export type LoadingState = 'PRE_SHOW' | 'SHOW_LOADING' | 'DONE'
+type LoadingStateRecord = Record<
+  string,
+  {
+    state: LoadingState
+    page: React.ComponentType<unknown>
+  }
+>
+
+interface ActivePageState {
+  loadingState: LoadingStateRecord
+}
+
+const ActivePageContext = React.createContext<ActivePageState | undefined>(
+  undefined
+)
+
+export const useActivePageContext = () => {
+  const activePageContext = useContext(ActivePageContext)
+
+  if (!activePageContext) {
+    throw new Error(
+      'useActivePageContext must be used within a ActivePageContext provider'
+    )
+  }
+
+  return activePageContext
+}
+
+type synchronousLoaderSpec = () => { default: React.ComponentType<unknown> }
+
+interface Props {
+  path: string
+  spec: Spec
+  delay?: number
+  params?: Record<string, string>
+  whileLoadingPage?: () => React.ReactElement | null
+  children?: React.ReactNode
+}
+
+const ArlNullPage = () => null
+const ArlWhileLoadingNullPage = () => null
+
+export const ActiveRouteLoader = ({
+  path,
+  spec,
+  delay,
+  params,
+  whileLoadingPage,
+  children,
+}: Props) => {
+  const location = useLocation()
+  const [pageName, setPageName] = useState('')
+  const loadingTimeout = useRef<NodeJS.Timeout>()
+  const announcementRef = useRef<HTMLDivElement>(null)
+  const mounted = useRef(true)
+  const waitingFor = useRef<string>('')
+  const [loadingState, setLoadingState] = useState<LoadingStateRecord>({
+    [path]: { page: ArlNullPage, state: 'PRE_SHOW' },
+  })
+  const [renderedChildren, setRenderedChildren] = useState<
+    React.ReactNode | undefined
+  >(children)
+  const [renderedPath, setRenderedPath] = useState(path)
+  const [renderedLocation, setRenderedLocation] = useState({ ...location })
+
+  loadingState[path] = loadingState[path] || 'PRE_SHOW'
+
+  const clearLoadingTimeout = () => {
+    if (loadingTimeout.current) {
+      clearTimeout(loadingTimeout.current)
+    }
+  }
+
+  useEffect(() => {
+    mounted.current = true
+
+    return () => {
+      mounted.current = false
+    }
+  }, [])
```

Comment:
I feel like we could get a few readability wins if we start extracting some of the logic here into hooks. For this one, something like:

```js
const useIsMounted = () => {
  const isMounted = useRef(false)

  useEffect(() => {
    isMounted.value = true
    return () => {
      isMounted.value = false
    }
  }, [])

  return isMounted.value
}
```

Reply:
You mean `isMounted.current`, right? (Not `value`)

Reply:
Right! I was doing some vue stuff a bit the other day and I think they have refs with `value` properties.

Reply:
Done

Reply:
I chose to put the hook in a separate file right next to this file rather than in some more globally accessible place because it's not very often you have to do something like this. Usually there are better way to handle your problems. So didn't want to make it too easy to reuse this hook 🙂

Reply:
Using your hook looks like this, right?
`const isMounted = useIsMounted()`

Now, if you use `isMounted` inside a `useEffect` you'll have to put `isMounted` in the dependency array. And that means when `isMounted` changes it'll trigger a rerender. And we don't want refs to trigger rerenders.

So what we have to do is return a function instead: `return useCallback(() => isMounted.current, [])`. Now `isMounted` is a stable value that's safe to put in dependency arrays. Just also have to change the usage to `isMounted()`